dialog/demo/demo.ts

```diff
@@ -11,10 +11,11 @@ import {KnobTypesToKnobs, MaterialCollection, materialInitsToStoryInits, setUpDe
 import {boolInput, Knob, selectDropdown, textInput} from './index.js';
 
 import {stories, StoryKnobs} from './stories.js';
+import { DialogTransition } from '../dialog.js';
 
 const collection =
     new MaterialCollection<KnobTypesToKnobs<StoryKnobs>>('Dialog', [
-      new Knob('transition', {
+      new Knob<Extract<DialogTransition, string>, "transition">('transition', {
```

Comment:
Can you check that out @asyncLiz @e111077 ?

That was the only way I could get rid of the error because `KnobTypesToKnobs` *extracts* the values of the generic so `DialogTransition` was no longer an option.
Maybe it would be better to change `KnobTypesToKnobs` directly? (I'm not a TypeScript wizard :disappointed_relieved: )

Reply:
I just confirm that this explicit generic makes TypeScript use types in proper places. It just doesn't feel right having to explicitly specify that.

Reply:
The thing is internally we actually do hook it up to the build system, but externally, even in the catalog, it is run in the playground which does not run any build processes. I might have to consider hooking this into the extenal build system as well. In the meantime I think there are 2 options:

1. follow the typing examples in [fab/demo/demo.ts](https://github.com/material-components/material-web/blob/main/fab/demo/demo.ts#L21-L29) and pass the generics and cast the default value
2. just skip the stories part and we'll do this for you since it just might be easier to do that ourselves

Reply:
@e111077 Ok I just updated , I understand now.

Reply:
A bit of context, is that Lit stories internally is framework agnostic (lit / angular / whatever) and is meant to be in a single file where the types are inferred from the `Knob()` array. We wanted to externalize these stories but only expose the simplest of stuff we could to the viewers of the catalog, so we had to do KnobTypesToKnobs shenanigans to format it in a way such that stories.ts is simple and the "source of truth", but still integrates with the internal build of our Lit stories system.

This is why there is some weirdness between that type and the `new Knob()` array.

Reply:
TBH I think this requires a rewrite of `selectDropdown`, but so many projects at google use lit stories, that i've been putting off that migration

Reply:
@e111077 Thanks a lot for the explanation, it means a lot. Definitely not easy to grasp some concept because I'm not familiar with the framework plus that stories TypeScript is crazy. But I see how generics work now.

I actually have a bigger question, what's your TS workflow for the demos? I manually copied the content of `stories` from the catalog to the dialog `demo` directory because the demo file was trying to import these files relatively..

Reply:
@e111077 Also could it be something is wrong with TypeScript from too much complexity? `KnobTypesToKnobs<StoryKnobs>` gives a decent list of knob types but somehow `DialogTransition` is lost in the `MaterialCollection` class template...
So writing
```typescript
new Knob<DialogTransition, "transition">('...
```
back get rid of the error.

Reply:
I agree that there are too many complex types, but that mostly stems from the complex types inside of Stories itself, which, in order to keep the same interface between internal and external build, might be difficult to simplify.

I feel the `KnobTypesToKnobs` complexity mostly stems from the complexity of stories itself and the attempt to keep it typesafe between stories.ts and demo.ts in each demo.

If you have any suggestions to change KnobTypesToKnobs to be simpler, i'd be very interested, but modifying the types inside Stories itself e.g. `Knob`, `Collection`, `LitCollection`, or `PolymorphicArrayOfKnobs`, there might be much more to consider

Reply:
I think the `KnobTypesToKnobs` type is fine, but for some reason I can't find a solution to that problem, the problem is that if there are string type values in the story the type string will get mixed with string union types (`DialogTransition` in this case) which causes the type to be lost in translation. I made a TypeScript playground : https://tsplay.dev/NBQMnw
 Maybe that can help you to find a solution?

Reply:
oh and nevermind, this will actually help intellisense but still getting some errors.
The only suggestion I would make is to export the `KnobInit` type so we can write:

```typescript
new Knob('transition', {
  defaultValue: 'grow-down',
  ui: selectDropdown({
    options: [
      {label: 'grow-down', value: 'grow-down'},
      // ...
    ]
  })
} as KnobInit<DialogTransition>),
```
only to avoid a double cast. (or else it's fine)